Web Inspector: Docked inspector should adjust according to the page's obscured content insets

the-chenergy opened the pull request:
#### e89594617c3cd8ca921e2164dd0d64a0498567b6
<pre>
Web Inspector: Docked inspector should adjust according to the page&apos;s obscured content insets
<a href="https://bugs.webkit.org/show_bug.cgi?id=291241">https://bugs.webkit.org/show_bug.cgi?id=291241</a>
<a href="https://rdar.apple.com/144717466">rdar://144717466</a>

Reviewed by Patrick Angle.

To respond to the host browser setting the _obscuredContentInsets of
an inspected WKWebView, the specialized WKInspectorViewController needs
to update the _obscuredContentInsets for the attached inspector, using
safeAreaInsets automatically calculated natively.

Using safeAreaInsets as the benchmark is sufficient because it takes
into account any custom insets applied on the inspector view&apos;s parent.

* Source/WebKit/UIProcess/Inspector/mac/WKInspectorViewController.mm:
(-[WKInspectorViewController webView]):
(-[WKInspectorViewController observeValueForKeyPath:ofObject:change:context:]):
(-[WKInspectorViewController webViewWebContentProcessDidTerminate:]):

Canonical link: <a href="https://commits.webkit.org/293770@main">https://commits.webkit.org/293770@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/930a29c08ed16ab9d36128ae143d04f6f8a79b58

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99894 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19542 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9826 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105021 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/50475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19847 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27977 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/76043 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/50475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102901 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/15146 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/90215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56401 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/14950 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49844 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/84877 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/8289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107382 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27007 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/19733 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/84997 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27369 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/86415 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84520 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29197 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/6924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/20826 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16245 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26944 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/32160 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/26755 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30071 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/28314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -3551,6 +3555,10 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
 #if PLATFORM(MAC)
+    // When there's an attached inspector, we compute and redistribute the client-set obscured content insets
+    // to the underlying page and the inspector. This field records the original client-set insets.
+    WebCore::FloatBoxExtent m_perceivedObscuredContentInsets;
```
In that case, I think this should be called something like `m_obscuredContentInsetsFromClient` to clarify that this value comes directly from the client (the "perceived" in here makes it sound more like the "effective" insets perceived by the user).

(I think you can also remove this comment, if you rename this member)

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -1933,7 +1933,7 @@ - (void)_recalculateViewportSizesWithMinimumViewportInset:(CocoaEdgeInsets)minim
 
 #if PLATFORM(MAC)
-    auto additionalInsets = _impl->obscuredContentInsets();
+    auto additionalInsets = _impl->effectiveObscuredContentInsetsConsideringAttachedInspector();
```
```suggestion
    auto additionalInsets = _impl->effectiveObscuredContentInsetsConsideringAttachedInspector();
```

I think that either `effectiveObscuredContentInsets` would be sufficient, or `obscuredContentInsetsAdjustedForAttachedInspector`.

whsieh commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -11543,7 +11552,11 @@ WebPageCreationParameters WebPageProxy::creationParameters(WebProcessProxy& proc
     parameters.textZoomFactor = m_textZoomFactor;
     parameters.pageZoomFactor = m_pageZoomFactor;
+#if PLATFORM(MAC)
+    parameters.obscuredContentInsets = perceivedObscuredContentInsets();
```
Hm.. is there a reason we propagate the client-specified value instead of the effective value to the newly created page?

the-chenergy replied:
Possibly not. I guess I just wasn't sure if "creation" parameters were meant to trace back the parameters that would be used to recreate the `WebPageProxy`.

patrickangle commented on Source/WebKit/UIProcess/Inspector/mac/WKInspectorViewController.mm:
```diff
@@ -57,6 +57,9 @@
 static NSString * const WKInspectorResourceScheme = @"inspector-resource";
 
+static NSString * const safeAreaInsetsKey = @"safeAreaInsets";
+static void* const webViewKvoSafeAreaInsetsKey = (void*)&webViewKvoSafeAreaInsetsKey;
```
Is this really a "key" or should this be called a "context" to match its usage below?

patrickangle replied:
Also nit: KVO is an acronym so should be all-caps in this case (i.e. `webViewKVOSafeAreaInsetsContext`)

patrickangle commented on Source/WebKit/UIProcess/Inspector/mac/WKInspectorViewController.mm:
```diff
@@ -110,11 +113,22 @@ - (WKWebView *)webView
         [_webView _setUseSystemAppearance:YES];
         [_webView setAutoresizingMask:NSViewWidthSizable | NSViewHeightSizable];
+
+        [_webView _setObscuredContentInsets:self.webView.safeAreaInsets immediate:NO];
```
Are you sure this is safe to do? Clients can set other insets beyond the safe area insets, for example if they are going to show a banner over part of the web view. In those cases it seems like the client's inset still needs to be honored...

the-chenergy replied:
I think this is quite safe to do. A view's `safeAreaInsets` should already honor any custom insets applied on its parent, including any `additionalSafeAreaInsets` or constraints it has. Sure, `safeAreaInsets` doesn't account for customizations on the view itself, but we as the owner of the inspector view know that it doesn't have any custom insets on it.

patrickangle replied:
We chatted offline. My key misunderstanding was forgetting that `_webView` is the Web Inspector's web view, not the web view that a client like Safari created.

the-chenergy committed:
Web Inspector: Docked inspector should adjust according to the page's obscured content insets
https://bugs.webkit.org/show_bug.cgi?id=291241
rdar://144717466

Reviewed by Patrick Angle.

To respond to the host browser setting the _obscuredContentInsets of
an inspected WKWebView, the specialized WKInspectorViewController needs
to update the _obscuredContentInsets for the attached inspector, using
safeAreaInsets automatically calculated natively.

Using safeAreaInsets as the benchmark is sufficient because it takes
into account any custom insets applied on the inspector view's parent.

* Source/WebKit/UIProcess/Inspector/mac/WKInspectorViewController.mm:
(-[WKInspectorViewController webView]):
(-[WKInspectorViewController observeValueForKeyPath:ofObject:change:context:]):
(-[WKInspectorViewController webViewWebContentProcessDidTerminate:]):

Canonical link: https://commits.webkit.org/293770@main
